File: rust/knn.rs
use ndarray::{Array2, Array1};
use ndarray::prelude::*;
use std::cmp::Ordering;

pub struct ClasificadorKVecinos {
    num_vecinos: usize,
    metrica: String,
    x_entrenamiento: Option<Array2<f64>>,
    y_entrenamiento: Option<Array1<usize>>,
}

impl ClasificadorKVecinos {
    // Crea una nueva instancia del clasificador k-NN
    pub fn nuevo(num_vecinos: usize, metrica: &str) -> Self {
        ClasificadorKVecinos {
            num_vecinos,
            metrica: String::from(metrica),
            x_entrenamiento: None,
            y_entrenamiento: None,
        }
    }

    // Ajusta el clasificador con los datos de entrenamiento y sus respectivas etiquetas
    pub fn ajustar(&mut self, datos_entrenamiento: &Array2<f64>, etiquetas_entrenamiento: &Array1<usize>) {
        self.x_entrenamiento = Some(datos_entrenamiento.clone());
        self.y_entrenamiento = Some(etiquetas_entrenamiento.clone());
    }

    // Calcula la precisión del clasificador utilizando datos de prueba y sus respectivas etiquetas
    pub fn puntuacion(&self, datos_prueba: &Array2<f64>, etiquetas_prueba: &Array1<usize>) -> f64 {
        let y_prediccion = self.predecir(datos_prueba);
        let precision = y_prediccion.iter().zip(etiquetas_prueba).filter(|&(x, y)| x == y).count() as f64 / y_prediccion.len() as f64;
        precision
    }

    // Predice las etiquetas de los datos de prueba utilizando el algoritmo k-NN
    pub fn predecir(&self, datos_prueba: &Array2<f64>) -> Array1<usize> {
        datos_prueba
            .axis_iter(Axis(0))
            .map(|x_prueba| {
                // Calcula las distancias entre el dato de prueba y los datos de entrenamiento
                let distancias = self.calcular_distancias(&x_prueba);
                
                // Ordena los índices de los datos de entrenamiento según las distancias
                let mut indices: Vec<usize> = (0..distancias.len()).collect();
                indices.sort_by(|a, b| distancias[*a].partial_cmp(&distancias[*b]).unwrap_or(Ordering::Equal));
                
                // Selecciona las etiquetas de los k vecinos más cercanos
                let etiquetas_vecinos: Vec<usize> = indices
                    .iter()
                    .take(self.num_vecinos)
                    .map(|&idx| self.y_entrenamiento.as_ref().unwrap()[idx])
                    .collect();
                
                // Encuentra la etiqueta más común entre los vecinos y la asigna como predicción
                let (etiqueta_predicha, _) = etiquetas_vecinos.into_iter().fold(std::collections::HashMap::new(), |mut acum, x| {
                    *acum.entry(x).or_insert(0) += 1;
                    acum
                }).into_iter().max_by_key(|&(_, conteo)| conteo).unwrap();
                etiqueta_predicha
            })
            .collect()
    }

    // Calcula las distancias entre un punto de prueba y todos los puntos de entrenamiento utilizando la métrica especificada
    fn calcular_distancias(&self, x: &Array1<f64>) -> Array1<f64> {
        self.x_entrenamiento.as_ref().unwrap().axis_iter(Axis(0)).map(|x_entrenamiento_i| {
            match self.metrica.as_str() {
                "euclidean" => x_entrenamiento_i.sq_l2_dist(&x).unwrap().sqrt(),
                "manhattan" => x_entrenamiento_i.l1_dist(&x).unwrap(),
                "minkowski" => {
                    let p = 3;
                    x_entrenamiento_i.iter().zip(x.iter()).map(|(&a, &b)| (a - b).abs().powi(p)).sum::<f64>().powf(1.0 / p as f64)
                }
                "chebyshev" => x_entrenamiento_i.iter().zip(x.iter()).map(|(&a, &b)| (a - b).abs()).fold(f64::MIN, f64::max),
                _ => panic!("Métrica no soportada: {}", self.metrica),
            }
        }).collect()
    }
}

File: python/knn.py
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
import os
from colorama import Fore
import pandas as pd

class MyKNeighborsClassifier:
    """
    Clasificador k-vecinos más cercanos personalizado.
    """

    def __init__(self, n_neighbors, metric):
        """
        Inicializa el clasificador.

        Parámetros:
        - n_neighbors: número de vecinos a considerar.
        - metric: métrica de distancia a utilizar ('euclidean' o 'manhattan').
        """
        self.n_neighbors = n_neighbors
        self.metric = metric
        self.X_train = None
        self.y_train = None

    def fit(self, datos_entrenamiento, etiquetas_entrenamiento):
        """
        Ajusta el modelo a los datos de entrenamiento.

        Parámetros:
        - datos_entrenamiento: matriz de características de los datos de entrenamiento.
        - etiquetas_entrenamiento: etiquetas de los datos de entrenamiento.
        """
        self.X_train = np.array(datos_entrenamiento)
        self.y_train = np.array(etiquetas_entrenamiento)

    def score(self, datos_prueba, etiquetas_prueba):
        """
        Calcula la precisión del modelo en los datos de prueba.

        Parámetros:
        - datos_prueba: matriz de características de los datos de prueba.
        - etiquetas_prueba: etiquetas de los datos de prueba.

        Retorna:
        - accuracy: precisión del modelo en los datos de prueba.
        """
        X_test = np.array(datos_prueba)
        y_test = np.array(etiquetas_prueba)
        y_pred = self.predict(X_test)
        accuracy = np.mean(y_pred == y_test)
        return accuracy

    def predict(self, datos_prueba):
        """
        Realiza predicciones para los datos de prueba.

        Parámetros:
        - datos_prueba: matriz de características de los datos de prueba.

        Retorna:
        - y_pred: etiquetas predichas para los datos de prueba.
        """
        X_test = np.array(datos_prueba)  # Convierte los datos de prueba a una matriz NumPy
        y_pred = []  # Lista para almacenar las etiquetas predichas

        # Itera sobre cada dato de prueba
        for i in range(X_test.shape[0]):
            distances = self.calculate_distances(X_test[i])  # Calcula las distancias entre el punto de prueba y los puntos de entrenamiento
            indices = np.argsort(distances)[:self.n_neighbors]  # Obtiene los índices de los vecinos más cercanos
            neighbors_labels = self.y_train[indices]  # Obtiene las etiquetas de los vecinos más cercanos

            # Cuenta las ocurrencias de cada etiqueta
            unique_labels, counts = np.unique(neighbors_labels, return_counts=True)

            # Encuentra la etiqueta más frecuente entre los vecinos más cercanos
            predicted_label = unique_labels[np.argmax(counts)]

            # Agrega la etiqueta predicha a la lista
            y_pred.append(predicted_label)

        return np.array(y_pred)  # Convierte la lista en un arreglo NumPy y lo retorna


    def kneighbors(self, datos, n_neighbors=1, return_distance=False):
        """
        Encuentra los k vecinos más cercanos para los datos dados.

        Parámetros:
        - datos: matriz de características de los datos.
        - n_neighbors: número de vecinos a considerar (por defecto es 1).
        - return_distance: indica si se deben devolver las distancias también (por defecto es False).

        Retorna:
        - Si return_distance es True:
            - distances: matriz de distancias a los k vecinos más cercanos.
        - Si return_distance es False:
            - indices: matriz de índices de los k vecinos más cercanos.
        """
        X = np.array(datos)  # Convierte los datos a una matriz NumPy
        distances = []  # Lista para almacenar las distancias
        indices = []  # Lista para almacenar los índices de los vecinos

        # Itera sobre cada dato
        for i in range(X.shape[0]):
            distances_i = self.calculate_distances(X[i])  # Calcula las distancias entre el dato y los puntos de entrenamiento
            indices_i = np.argsort(distances_i)[:n_neighbors]  # Obtiene los índices de los vecinos más cercanos

            distances.append(distances_i[indices_i])  # Agrega las distancias de los vecinos más cercanos
            indices.append(indices_i)  # Agrega los índices de los vecinos más cercanos

        if return_distance:
            return np.array(distances)  # Retorna una matriz de distancias
        else:
            return np.array(indices)  # Retorna una matriz de índices


    def calculate_distances(self, x):
        """
        Calcula las distancias entre un punto dado y los puntos de entrenamiento.

        Parámetros:
        - x: punto para el cual se calcularán las distancias.
        
        Retorna:
        - distances: matriz de distancias entre el punto dado y los puntos de entrenamiento.
        """

        distances = []
        for i in range(self.X_train.shape[0]):
            if self.metric == 'euclidean':
                print(self.X_train[i])
                distance = np.sqrt(np.sum((self.X_train[i] - x) ** 2))  # Distancia Euclidiana
            elif self.metric == 'manhattan':
                distance = np.sum(np.abs(self.X_train[i] - x))  # Distancia Manhattan
            elif self.metric == 'minkowski':
                p = 3  # Parámetro p para la distancia Minkowski (puedes ajustar su valor según necesites)
                distance = np.power(np.sum(np.power(np.abs(self.X_train[i] - x), p)), 1/p)  # Distancia Minkowski
            elif self.metric == 'chebyshev':
                distance = np.max(np.abs(self.X_train[i] - x))  # Distancia Chebyshev
            else:
                raise ValueError("Unsupported metric: {}".format(self.metric))

            distances.append(distance)

        return np.array(distances)


def getKnn(datos_entrenamiento, etiquetas_entrenamiento, k, metric='manhattan', model="myModel"):
    # Creamos el objeto clasificador KNN con el valor de k especificado

    if model == "myModel" :
        knn = MyKNeighborsClassifier(n_neighbors=k, metric=metric)
        # Entrenamos el clasificador KNN con los datos de entrenamiento
        knn.fit(datos_entrenamiento, etiquetas_entrenamiento)
        #devolvemos el modelo ya entrenado
        return knn
    elif model == "librery":
        knn = KNeighborsClassifier(n_neighbors=k, metric=metric)
        # Entrenamos el clasificador KNN con los datos de entrenamiento
        knn.fit(datos_entrenamiento, etiquetas_entrenamiento)
        #devolvemos el modelo ya entrenado
        return knn
    
    raise ValueError("llegó aquí por ",model)




def export_to_csv(data, filename):
    """
    Exporta un objeto pandas.DataFrame a un archivo CSV con el nombre especificado.
    """
    data.to_csv(filename, index=False)

def import_from_csv(filename):
    """
    Importa un archivo CSV y retorna un objeto pandas.DataFrame con su contenido.
    """
    try:
        data = pd.read_csv(filename)
    except FileNotFoundError:
        print(Fore.RED + f"Error: El archivo {filename} no existe.")
        exit(1)
    except pd.errors.ParserError:
        print(Fore.RED + f"Error: El archivo {filename} no tiene un formato válido.")
        exit(1)

    if data is None:
        print(Fore.RED + "Error: El archivo {filename} está vacio.")
        exit(1)

    return data

def getDataSets():
    path= os.path.abspath("./dataset")
    datasets = {}

    if not os.listdir(path):
        print(Fore.RED + "Error: No hay ningún dataset para cargar." )
        exit(1)

    for foldername in os.listdir(path):
        folderpath = os.path.join(path, foldername)
        if os.path.isdir(folderpath):
            data = import_from_csv(folderpath + '/data.csv')
            target =  import_from_csv(folderpath + '/target.csv')
            datasets[foldername] = (data, target)
    return datasets


# Ejemplo de uso con el dataset Iris
#from sklearn.datasets import load_iris

#iris = load_iris()

# Exportar iris.target a un archivo CSV
#export_to_csv(pd.DataFrame(iris.target), "iris_target.csv")

# Exportar iris.data a un archivo CSV
#export_to_csv(pd.DataFrame(iris.data), "iris_data.csv")

# Importar iris.target desde un archivo CSV
#target = import_from_csv("iris_target.csv")

# Importar iris.data desde un archivo CSV
#data = import_from_csv("iris_data.csv")




